Content.Server/Storage/Components/ServerStorageComponent.cs

```diff
@@ -134,7 +137,23 @@ public bool CanInsert(IEntity entity)
                 return false;
             }
 
-            if (_whitelist != null && !_whitelist.IsValid(entity))
+            // Item is not in whitelist & there is no max item size.
+            if (_whitelist != null && !_whitelist.IsValid(entity) && (_storageMaxItemSize == null))
+            {
+                return false;
+            }
+
+            // Item is above max item size
+            if (_storageMaxItemSize != null && store?.Size > _storageMaxItemSize)
+            {
+                // Item is not whitelisted or there isn't a whitelist.
+                if (_whitelist == null || (_whitelist != null && !_whitelist.IsValid(entity)))
+                {
+                    return false;
+                }
+            }
```

Comment:
i dont get this part:
why not just
```suggestion
            // Item is not in whitelist
            if (_whitelist != null && !_whitelist.IsValid(entity))
            {
                return false;
            }

            // Item is above max item size
            if (_storageMaxItemSize != null && store?.Size > _storageMaxItemSize)
            {
                return false;
            }
```

Reply:
Such solution would seem optimal at first, but it brings undesired behaviour.
It makes so that having a whitelist AND a max item size is pointless.

**Example:**
- Storage A has 40 capacity, and a max item size of 5.
- Item X occupies 10 of space.
- Item Y occupies 15 of space.
- Item Z occupies  5  of space.

Normally, Item X and Y do not fit into storage A due to its size.
However if a whitelist is applied containing the items we want to fit, the max size becomes pointless. And if the max size can dictate if something is rejected, if Item X is whitelisted, it will still not fit because of the max item size.

Instead, when using a whitelist AND a max item size, it is expected that all items that fit within the max item size to be allowed, and if a item does not match the max item size but is in the whitelist, it is also allowed.

The whitelist together with the max item size serves as a exception list to the max item size allowing further balancing options.

Reply:
> i dont get this part:
> why not just
> ```suggestion
>             // Item is not in whitelist
>             if (_whitelist != null && !_whitelist.IsValid(entity))
>             {
>                 return false;
>             }
> 
>             // Item is above max item size
>             if (_storageMaxItemSize != null && store?.Size > _storageMaxItemSize)
>             {
>                 return false;
>             }
> ```

Such solution would seem optimal at first, but it brings undesired behaviour.
It makes so that having a whitelist AND a max item size is pointless.

**Example:**
- Storage A has 40 capacity, and a max item size of 5.
- Item X occupies 10 of space.
- Item Y occupies 15 of space.
- Item Z occupies  5  of space.

Normally, Item X and Y do not fit into storage A due to its size.
However if a whitelist is applied containing the items we want to fit, the max size becomes pointless. And if the max size can dictate if something is rejected, if Item X is whitelisted, it will still not fit because of the max item size.

Instead, when using a whitelist AND a max item size, it is expected that all items that fit within the max item size to be allowed, and if a item does not match the max item size but is in the whitelist, it is also allowed.

The whitelist together with the max item size serves as a exception list to the max item size allowing further balancing options.

Reply:
what you are thinking off would be a "bypasslist" of sorts. thats not what the whitelist has been so far and i dont think we should change that. I'm not against adding a bypasslist tho

Reply:
The whitelist behaviour isn't changing if there is not a max item size, which is why its a nullable.

Reply:
> 
> 
> The whitelist behaviour isn't changing if there is not a max item size, which is why its a nullable.

but it will change behaviour if the max size is specified, hence my point.

i'll just say it bluntly, i wont merge it with the current behaviour. i'm not gonna block other maintainers from doing so, but i doubt other will have different opinions than mine.

Reply:
I agree with Paul; every check should be passed or insertion should fail.

Reply:
Conceptually, your proposal does an OR between the max_item_size check and whitelist check. So `fits = (max_item_size || whitelist)`. 

After a long and furious battle on Discord we came to the agreement that having a separate "bypass" list would be better. This is because an AND'd whitelist may still be useful *on top* of the proposed behavior (since it works with tags and components). 

So the logic would look like this: `fits = whitelist && (max_item_size || bypass_list)`. This would allow you to express "all tools below size 10 and also fits the jaws of life"

SS13 also has separate properties for this but the logic seems to make no sense to me (if `can_only_hold` then max item size is completely ignored, but `fits_ignoring_w_class` is union'd to `can_only_hold` which ???) similar: https://github.com/vgstation-coders/vgstation13/blob/2ded66406d928e1e2ff9f06a4253cd984ce1f00a/code/game/objects/items/weapons/storage/storage.dm#L13-L21